### ISSUE TITLE
UB in C++ but not UB in C: Right Shift Bug

### ISSUE THREAD

**Issue by z3rodae0** [emscripten-core/emscripten#20481]
### Version of emscripten/emsdk:
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.47 (431685f05c67f0424c11473cc16798b9587bb536)
clang version 18.0.0 (https://github.com/llvm/llvm-project 21030b9ab4487d845e29792063f5666d8c4b8e09)
Target: wasm32-unknown-emscripten
Thread model: posix
InstalledDir: /home/z3rodae0/emscripten/upstream/bin

### Collection of Code Where Bugs Occur:
[binarys.zip](https://github.com/emscripten-core/emscripten/files/12985445/binarys.zip)
The binarys.zip file contains problematic source files: driver.c, func.c, init.h, and corresponding compile.sh and execution.sh scripts for compiling and running them with various optimization options.

### Excution Results:
```
[+] node : emcc_O0.js
warning: running JS from STANDALONE_WASM without WASM_BIGINT will fail if a syscall with i64 is used (in standalone mode we cannot legalize syscalls)
2
[+] node : emcc_O1.js
0
[+] node : emcc_O2.js
0
[+] node : emcc_O3.js
0
[+] wasmer : emcc_O0.wasm
2
[+] wasmer : emcc_O1.wasm
0
[+] wasmer : emcc_O2.wasm
0
[+] wasmer : emcc_O3.wasm
0
[+] wasmtime : emcc_O0.wasm
2
[+] wasmtime : emcc_O1.wasm
0
[+] wasmtime : emcc_O2.wasm
0
[+] wasmtime : emcc_O3.wasm
0
```
The issue could be with the compiler or might also occur in the WebAssembly runtime tools, so I tested it with multiple CLI runtime tools.

### RootCause
```
1 // func.c
2 #include "init.h"
3 void tf_1_foo ()
4 {
5     long long int a = 0;
6     result = (2 >> ((a) >> (64)));
7 }
```

When a bitshift operation is performed on line 6 with a shift size greater than the variable's size declared on line 5, a bug occurs. In the context of C++, this behavior is considered undefined, but in plain C, it is not. In the C compiler emcc, the expected behavior is to output the value 2. Other C compilers consistently output 2 for each optimization option. However, emcc, when optimization is enabled, outputs 0 instead of the expected 2.

**Comment by sbc100** [comment]
Are you sure about "Other C compilers consistently output 2 for each optimization option"?

I get the same results (either 0 or 2 depending on the opt level) when I use clang.  At least for this simplified version of the program:

```
$ cat test.c
#include <stdio.h>

int result;

void tf_1_foo() {
  long long int a = 0;
  result = (2 >> ((a) >> (64)));
}

int main() {
  tf_1_foo();
  printf("%d\n", result);
}
$ clang -O0  test.c && ./a.out
2
$ clang -O1 test.c && ./a.out
0
$ ./emcc -O0 test.c && node ./a.out.js
2
$ ./emcc -O1 test.c && node ./a.out.js
0
```

The version of `gcc` on my system however does always produce 2 even at high opt levels, so perhaps this is a clang/llvm thing?

**Comment by juj** [comment]
> In the context of C++, this behavior is considered undefined, but in plain C, it is not.

[ISO/IEC C99 Standard](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf) (oldest C standard available for free):

<img width="640" alt="image" src="https://github.com/emscripten-core/emscripten/assets/225351/a2d9edda-e2e9-41c3-9395-fade66103b99">

_"If the value of the right operand is [...] equal to the width of the promoted left operand, the behavior is undefined."_

---------

[ISO/IEC C17 Standard](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2310.pdf) (newest C standard available for free):

<img width="625" alt="image" src="https://github.com/emscripten-core/emscripten/assets/225351/2a44cad4-1939-4326-8d95-4d25665e219d">

_"If the value of the right operand is [...] equal to the width of the promoted left operand, the behavior is undefined."_

---------

[ISO/IEC C++98 Standard](https://www.lirmm.fr/~ducour/Doc-objets/ISO+IEC+14882-1998.pdf) (oldest C++ standard available):

<img width="646" alt="image" src="https://github.com/emscripten-core/emscripten/assets/225351/16ff3d17-12ba-4b0b-a556-f9adeab355cf">

_"The behavior is undefined if the right operand is [...] equal to the length in bits of the promoted left operand."_

---------

[ISO/IEC C++20 Standard](https://isocpp.org/files/papers/N4860.pdf): (newest ratified C++ standard):

<img width="665" alt="image" src="https://github.com/emscripten-core/emscripten/assets/225351/10d546bb-37e4-403c-bca8-93e259616d30">

_" The behavior is undefined if the right operand is [...] equal to the width of the promoted left operand."_

---------

C Standards PDFs can be found at https://www.iso-9899.info/wiki/The_Standard.

C++ standards PDFs can be found at https://en.cppreference.com/w/cpp/links.

In `(a) >> (64)`, `a` is 64 bits in Emscripten, hence the expression would be undefined behavior both in old and new C and C++ standards, according to the above?

**Comment by z3rodae0** [comment]
First of all, thank you for your interest in my bug report. I was already somewhat aware that this behavior could be undefined, and now I understand that it is indeed UB in both C and C++. However, the reason I still consider this a bug is as follows:

The results vary depending on the optimization level, meaning that the behavior of UB is not consistent.
The behavior of this operation is correctly produced by the older and mature compiler, GCC, for different optimization levels.
In conclusion, the key issue in this problem is not UB but the varying output values across different optimization levels.

Of course, I'm not completely sure, as my knowledge is limited. I would appreciate any feedback you can provide.

**Comment by RReverser** [comment]
> The results vary depending on the optimization level, meaning that the behavior of UB is not consistent.

Well, that's pretty much the definition of UB - that it's undefined, will depend on arbitrary factors and you can't rely on any specific behaviour.

**Comment by z3rodae0** [comment]
Um... I see. I have another similar case. When I perform a left shift operation and bits get truncated, the result is binary 0. With clang and gcc at O0, O1, O2, and O3, it's consistently 0. However, in the binary compiled with emcc, the result is different. Please explain in the comments why it varies with different optimization levels. I don't know how to debug in WebAssembly (other than using Chrome Developer Tools, I prefer doing it in a CLI environment).

I think it might be due to highly optimized function calls

```
//driver.c
#include <stdio.h>
#include "init.h"
int bug = 1;

extern void tf_4_foo ();
int main () {
    func();

    printf("%llu\n", bug);
    return 0;
}
```
```
// func.c
#include "init.h"

void func ()
{
    bug =  ((0L) << 32);
}
```

```
// init.h
extern long int bug;
```

**Comment by juj** [comment]
Of additional note is that the LLVM compiler does correctly diagnose the scenario:

`b.cpp`
```c++
// func.c
void tf_1_foo ()
{
    long long int a = 0;
    result = (2 >> ((a) >> (64)));
}
```

```
b.cpp:5:25: warning: shift count >= width of type [-Wshift-count-overflow]
    5 |     result = (2 >> ((a) >> (64)));
      |                         ^  ~~~~
```

> The results vary depending on the optimization level, meaning that the behavior of UB is not consistent.
The behavior of this operation is correctly produced by the older and mature compiler, GCC, for different optimization levels.
In conclusion, the key issue in this problem is not UB but the varying output values across different optimization levels.

I can appreciate reasonings like this that give a desire to remove this type of source of UB.

However, that is a C/C++ standards question and not an Emscripten question. If you'd like a resolution here, you can try:
 - contact the C Standard committee at https://www.open-std.org/JTC1/SC22/WG14/www/contacts, or
 - contact the C++ Standard committee at https://isocpp.org/std/submit-a-proposal, or
 - if you think LLVM should do a compiler-specific behavior to override the standards, then contact them at https://github.com/llvm/llvm-project/issues

Any of those directions would probably be able to give a better rationale than I could of why exactly this kind of shifting behavior is UB in the standard. (I believe it boils down to performance-related concerns over hardware instruction set encoding).

I'll close this issue in the Emscripten bug tracker - we are after all a downstream users of the standards and the LLVM compiler.

**Comment by juj** [comment]
`(0L) << 32`

Emscripten is a 32-bit platform, where `sizeof(long)` is 4, so 32 bits. Other 64-bit targets may have `sizeof(long)` be 8, i.e. 64 bits, so on those platforms, `(0L) << 32` would be well defined, but with Emscripten with `long` only 32 bits wide, that falls under the UB scenario.
